internal/engine/planner.go

```diff
@@ -32,105 +31,152 @@ type (
 		node *qpN
 		role string
 	}
-	NoSuchKeyError struct {
-		msg string
+
+	PolicyPlanResult struct {
+		Scope       string
+		DenyFilter  []*qpN
+		AllowFilter []*qpN
 	}
 )
 
-func (e *NoSuchKeyError) Error() string {
-	return e.msg
+func combinePlans(principalPolicyPlan, resourcePolicyPlan *PolicyPlanResult) *PolicyPlanResult {
+	if principalPolicyPlan.Empty() {
+		return resourcePolicyPlan
+	}
+
+	if resourcePolicyPlan.Empty() {
+		return principalPolicyPlan
+	}
+
+	return &PolicyPlanResult{
+		Scope:       fmt.Sprintf("principal: %q; resource: %q", principalPolicyPlan.Scope, resourcePolicyPlan.Scope),
```

Comment:
I'm not 100% happy with packing information into a string like this rather than using a struct, but it's only used in the optional response metadata so it doesn't seem worth a breaking API change? 🤷‍♂️

Reply:
Aaargh! This is something I hadn't considered before. Usually there would be only one applicable scope but that changes here, doesn't it?

I think what you've done here is the best way to work around it. It's mostly for informational purposes anyway.